modules/ephemeris/src/main/scala/gem/horizons/HorizonsNameQuery.scala

```diff
@@ -0,0 +1,243 @@
+
+  /** URL string corresponding the the horizons request. */
+  def urlString: String
+
+  /** Returns a program that will perform the name lookup when executed,
+    * returing a List of ephemeris keys with their common name.
+    */
+  def lookup: Result[List[Resolution[A]]]
+}
+
+object HorizonsNameQuery {
+
+  /** Describes a non-sidereal target query of a specific type of object. */
+  sealed abstract class Search[A](val queryString: String) extends Product with Serializable
+
+  object Search {
+    final case class Comet(partial: String)     extends Search[EphemerisKey.Comet    ](s"NAME=$partial*;CAP")
+    final case class Asteroid(partial: String)  extends Search[EphemerisKey.Asteroid ](s"ASTNAM=$partial*"  )
+    final case class MajorBody(partial: String) extends Search[EphemerisKey.MajorBody](s"$partial"          )
+  }
+
+  sealed trait Error extends Product with Serializable
+
+  final case class HorizonsException(e: Throwable)                  extends Error
+  final case class ParseError(input: List[String], message: String) extends Error
+
+  /** Result of performing an horizons name resolution. */
+  type Result[A] = EitherT[IO, Error, A]
+
+  object Result {
+    def fromEither[A](e: Either[Error, A]): Result[A] =
+      EitherT.fromEither[IO](e)
+
+    def delay[A](a: => A): Result[A] =
+      EitherT.rightT[IO, Error](a)
+  }
+
+  /** A human-readable name and its unique horizons ephemeris key. */
+  final case class Resolution[A](a: A, name: String)
+
+  /** Creates a NameQuery for the given search object. */
+  def apply[A](search: Search[A]): HorizonsNameQuery[A] =
+    new HorizonsNameQuery[A] {
+      import NameQueryImpl._
+
+      val reqParams = HorizonsClient.SharedParams ++ Map(
+        "MAKE_EPHEM" -> "NO",
+        "COMMAND"    -> s"'${search.queryString}'"
+      )
+
+      override val urlString: String =
+        HorizonsClient.urlString(reqParams)
+
+      override val lookup: Result[List[Resolution[A]]] =
+        EitherT(HorizonsClient.fetch(reqParams).map { s =>
+          val lines = s.split('\n').toList
+          parseResponse(search, lines).leftMap(ParseError(lines, _))
+        })
+    }
+
+  private object NameQueryImpl {
+
+    // Representation of column offsets as (start, end) pairs, as deduced
+    // from --- -------- --- -----
+    type Offsets = List[(Int, Int)]
+
+    // Parse many results based on an expected header pattern. This will read
+    // through the tail until --- separators are found, then parse the
+    // remaining lines (until a blank line is encountered) using the function
+    // constructed by `f`, if any (otherwise it is assumed that there were not
+    // enough columns found). Returns a list of values or an error message.
+    def parseMany[A](header: String, tail: List[String], headerPattern: Regex)(f: Offsets => Option[String => A]): Either[String, List[A]] =
+      headerPattern.findFirstMatchIn(header).toRight("Header pattern not found: " + headerPattern.regex) *>
```

Comment:
This code makes me really nervous but it was the best I could do at the time. It would be nice to revisit this at some point and see if it can be made more obvious. Ok for now though.